deepchem/rl/a2c.py

```diff
@@ -1,23 +1,32 @@
 """Advantage Actor-Critic (A2C) algorithm for reinforcement learning."""
 
-from deepchem.models import KerasModel
-from deepchem.models.optimizers import Adam
 import numpy as np
-import tensorflow as tf
 import collections
 import copy
 import multiprocessing
 import os
 import re
 import threading
 import time
+from deepchem.models import KerasModel
+from deepchem.models.optimizers import Adam
 
 
 class A2CLossDiscrete(object):
-  """This class computes the loss function for A2C with discrete action spaces."""
+  """This class computes the loss function for A2C with discrete action spaces.
+
+  Note
+  ----
+  This class requires tensorflow to be installed.
+  """
 
   def __init__(self, value_weight, entropy_weight, action_prob_index,
                value_index):
+
```

Comment:
One question this PR raises for me is whether we should have class level import guards in addition to module level import guards. I think I'm tentatively in favor of having both since it keeps classes self-contained in case someone tries to import a class in directly and use it, but I'm open to taking them out as well.

What do you folks think? CC @vsomnath @peastman

Reply:
Alternatively we could go in the other direction.  Have the files for individual classes (e.g. `a2c.py`) go ahead and `import tensorflow` at the top level.  The package level import guards could then be simplified to

```python

try:
  from deepchem.rl.a2c import A2C
  from deepchem.rl.ppo import PPO
except ModuleNotFoundError:
  class A2C(TensorFlowStub):
    pass
  class PPO(TensorFlowStub):
    pass
```

You would then get this behavior:

- If you `import deepchem as dc`, you can write `dc.rl.A2C` as usual, and you get the stub class.
- If you write `from deepchem.rl import A2C` it also works as expected, and you get the stub class.
- If you write `from deepchem.rl.a2c import A2C` you instead get an exception that `tensorflow` can't be imported.  That's still reasonably clear, but also is perhaps a reminder that you really shouldn't be importing it this way.